### ISSUE TITLE
[Web] WebGPU and WASM Backends Unavailable within Service Worker

### ISSUE THREAD

**Issue by ggaabe** [microsoft/onnxruntime#20876]
### Describe the issue

I'm running into issues trying to use the WebGPU or WASM backends inside of a ServiceWorker (on a chrome extension). More specifically, I'm attempting to use Phi-3 with transformers.js v3

Every time I attempt this, I get the following error:

```
Uncaught (in promise) Error: no available backend found. ERR: [webgpu] 
TypeError: import() is disallowed on ServiceWorkerGlobalScope by the HTML specification. 
See https://github.com/w3c/ServiceWorker/issues/1356.
```

This is originating in the `InferenceSession` class in `js/common/lib/inference-session-impl.ts`.

More specifically, it's happening in this method:
`const [backend, optionsWithValidatedEPs] = await resolveBackendAndExecutionProviders(options);`
where the implementation is in `js/common/lib/backend-impl.ts` and the `tryResolveAndInitializeBackend` fails to initialize any of the execution providers.

WebGPU is now supported in ServiceWorkers though; it is a recent change and it should be feasible. [Here were the chrome release notes](https://developer.chrome.com/blog/new-in-webgpu-124?hl=en#service_workers_and_shared_workers_support).

Additionally, here is an example browser extension from the mlc-ai/web-llm framework that implements WebGPU usage in service workers successfully:
https://github.com/mlc-ai/web-llm/tree/main/examples/chrome-extension-webgpu-service-worker

Here is some further discussion on this new support from Google itself:
https://groups.google.com/a/chromium.org/g/chromium-extensions/c/ZEcSLsjCw84/m/WkQa5LAHAQAJ

So technically I think it should be possible for this to be supported now? Unless I'm doing something else glaringly wrong. Is it possible to add support for this?

### To reproduce

Download and set up the transformers.js [extension example](https://github.com/xenova/transformers.js/tree/main/examples/extension) and put this into the background.js file:
```js
// background.js - Handles requests from the UI, runs the model, then sends back a response

import {
  pipeline,
  env,
  AutoModelForCausalLM,
  AutoTokenizer,
  TextStreamer,
  StoppingCriteria,
} from "@xenova/transformers";

// Skip initial check for local models, since we are not loading any local models.
env.allowLocalModels = false;

// Due to a bug in onnxruntime-web, we must disable multithreading for now.
// See https://github.com/microsoft/onnxruntime/issues/14445 for more information.
env.backends.onnx.wasm.numThreads = 1;

class CallbackTextStreamer extends TextStreamer {
  constructor(tokenizer, cb) {
    super(tokenizer, {
      skip_prompt: true,
      skip_special_tokens: true,
    });
    this.cb = cb;
  }

  on_finalized_text(text) {
    this.cb(text);
  }
}

class InterruptableStoppingCriteria extends StoppingCriteria {
  constructor() {
    super();
    this.interrupted = false;
  }

  interrupt() {
    this.interrupted = true;
  }

  reset() {
    this.interrupted = false;
  }

  _call(input_ids, scores) {
    return new Array(input_ids.length).fill(this.interrupted);
  }
}

const stopping_criteria = new InterruptableStoppingCriteria();

async function hasFp16() {
  try {
    const adapter = await navigator.gpu.requestAdapter();
    return adapter.features.has("shader-f16");
  } catch (e) {
    return false;
  }
}

class PipelineSingleton {
  static task = "feature-extraction";
  static model_id = "Xenova/Phi-3-mini-4k-instruct_fp16";
  static model = null;
  static instance = null;

  static async getInstance(progress_callback = null) {
    this.model_id ??= (await hasFp16())
      ? "Xenova/Phi-3-mini-4k-instruct_fp16"
      : "Xenova/Phi-3-mini-4k-instruct";

    this.tokenizer ??= AutoTokenizer.from_pretrained(this.model_id, {
      legacy: true,
      progress_callback,
    });

    this.model ??= AutoModelForCausalLM.from_pretrained(this.model_id, {
      dtype: "q4",
      device: "webgpu",
      use_external_data_format: true,
      progress_callback,
    });

    return Promise.all([this.tokenizer, this.model]);
  }
}

// Create generic classify function, which will be reused for the different types of events.
const classify = async (text) => {
  // Get the pipeline instance. This will load and build the model when run for the first time.
  const [tokenizer, model] = await PipelineSingleton.getInstance((data) => {
    // You can track the progress of the pipeline creation here.
    // e.g., you can send `data` back to the UI to indicate a progress bar
    console.log("progress", data);
    // data logs as this:
    /**
     * 
     * {
    "status": "progress",
    "name": "Xenova/Phi-3-mini-4k-instruct_fp16",
    "file": "onnx/model_q4.onnx",
    "progress": 99.80381792394503,
    "loaded": 836435968,
    "total": 838080131
  }

  when complete, last status will be 'done'
     */
  });
  /////////////
  const inputs = tokenizer.apply_chat_template(text, {
    add_generation_prompt: true,
    return_dict: true,
  });

  let startTime;
  let numTokens = 0;
  const cb = (output) => {
    startTime ??= performance.now();

    let tps;
    if (numTokens++ > 0) {
      tps = (numTokens / (performance.now() - startTime)) * 1000;
    }
    self.postMessage({
      status: "update",
      output,
      tps,
      numTokens,
    });
  };

  const streamer = new CallbackTextStreamer(tokenizer, cb);

  // Tell the main thread we are starting
  self.postMessage({ status: "start" });

  const outputs = await model.generate({
    ...inputs,
    max_new_tokens: 512,
    streamer,
    stopping_criteria,
  });
  const outputText = tokenizer.batch_decode(outputs, {
    skip_special_tokens: false,
  });

  // Send the output back to the main thread
  self.postMessage({
    status: "complete",
    output: outputText,
  });
  ///////////////

  // Actually run the model on the input text
  // let result = await model(text);
  // return result;
};

////////////////////// 1. Context Menus //////////////////////
//
// Add a listener to create the initial context menu items,
// context menu items only need to be created at runtime.onInstalled
chrome.runtime.onInstalled.addListener(function () {
  // Register a context menu item that will only show up for selection text.
  chrome.contextMenus.create({
    id: "classify-selection",
    title: 'Classify "%s"',
    contexts: ["selection"],
  });
});

// Perform inference when the user clicks a context menu
chrome.contextMenus.onClicked.addListener(async (info, tab) => {
  // Ignore context menu clicks that are not for classifications (or when there is no input)
  if (info.menuItemId !== "classify-selection" || !info.selectionText) return;

  // Perform classification on the selected text
  let result = await classify(info.selectionText);

  // Do something with the result
  chrome.scripting.executeScript({
    target: { tabId: tab.id }, // Run in the tab that the user clicked in
    args: [result], // The arguments to pass to the function
    function: (result) => {
      // The function to run
      // NOTE: This function is run in the context of the web page, meaning that `document` is available.
      console.log("result", result);
      console.log("document", document);
    },
  });
});
//////////////////////////////////////////////////////////////

////////////////////// 2. Message Events /////////////////////
//
// Listen for messages from the UI, process it, and send the result back.
chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
  console.log("sender", sender);
  if (message.action !== "classify") return; // Ignore messages that are not meant for classification.

  // Run model prediction asynchronously
  (async function () {
    // Perform classification
    let result = await classify(message.text);

    // Send response back to UI
    sendResponse(result);
  })();

  // return true to indicate we will send a response asynchronously
  // see https://stackoverflow.com/a/46628145 for more information
  return true;
});
```

### Urgency

this would help enable a new ecosystem to build up around locally intelligent browser extensions and tooling.

it's urgent for me because it would be fun to build and I want to build it and it would be fun to be building it rather than not be building it.

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

1.19.0-dev.20240509-69cfcba38a

### Execution Provider

'webgpu' (WebGPU)

**Comment by fs-eire** [comment]
Than you for reporting this issue. I will try to figure out how to fix this problem.

**Comment by fs-eire** [comment]
So it turns out to be that dynamic import (ie. `import()`) and top-level `await` is not supported in current service worker. I was not expecting 

Currently, the WebAssembly factory (wasm-factory.ts) uses dynamic import to load the JS glue. This does not work in service worker. A few potential solutions are also not available:
- Modifying it to import statement: won't work, because the JS glue includes top-level `await`.
- Using `importScripts`: won't work, because the JS glue is ESM
- Using `eval`: won't work; same to `importScripts`

I am now trying to make a JS bundle that does not use dynamic import for usage of service worker specifically. Still working on it

**Comment by ggaabe** [comment]
Thanks,  I appreciate your efforts around this. It does seem like some special-case bundle will need to be built after all; you might need `iife` or `umd` for the bundler output format

**Comment by fs-eire** [comment]
> Thanks, I appreciate your efforts around this. It does seem like some special-case bundle will need to be built after all; you might need `iife` or `umd` for the bundler output format

I have considered this option. However, Emscripten does not offer an option to output both UMD(IIFE+CJS) & ESM for JS glue (https://github.com/emscripten-core/emscripten/issues/21899). I have to choose either. I choose the ES6 format output for the JS glue, because of a couple of problems when import UMD from ESM, and `import()` is a standard way to import ESM from both ESM and UMD. ( Until I know its not working in service worker by this issue)

**Comment by fs-eire** [comment]
@ggaabe  Could you please help to try `import * as ort from “./ort.webgpu.bundle.min.js”` from version 1.19.0-dev.20240604-3dd6fcc089 ?

**Comment by ggaabe** [comment]
@fs-eire my project is dependent on transformersjs, which imports onnxruntime webgpu backend like this here:

https://github.com/xenova/transformers.js/blob/v3/src/backends/onnx.js#L24

Is this the right usage? In my project I've added this to my package.json to resolve onnx-runtime to this new version though the issue is still occurring:

```
  "overrides": {
    "onnxruntime-web": "1.19.0-dev.20240604-3dd6fcc089"
  }
```

**Comment by ggaabe** [comment]
Maybe also important: The same error is still occurring in same spot in inference session in the onnx package and not from transformersjs. Do I need to add a resolver for onnxruntime-common as well?

**Comment by fs-eire** [comment]
#20991 makes default ESM import to use non-dynamic-import and hope this change may fix this problem. PR is still in progress

**Comment by ggaabe** [comment]
Hi @fs-eire, is the newly-merged fix in a released build I can try?

**Comment by fs-eire** [comment]
Please try 1.19.0-dev.20240612-94aa21c3dd

**Comment by ggaabe** [comment]
@fs-eire there's an error indicated from `//# sourceMappingURL=ort.webgpu.bundle.min.mjs.map` because it tries referencing a document (documents aren't available in service workers)
`Uncaught ReferenceError: document is not defined`
```
/******/ 	/* webpack/runtime/jsonp chunk loading */
/******/ 	(() => {
/******/ 		__webpack_require__.b = document.baseURI || self.location.href;
````

It looks like this source code is the likely culprit of what we see getting transpiled into the above cause of the error:

https://github.com/microsoft/onnxruntime/blob/main/js/web/lib/wasm/wasm-utils-import.ts#L12-L20

```ts
export const scriptSrc =
    // if Nodejs, return undefined
    isNode ? undefined :
             // if It's ESM, use import.meta.url
             BUILD_DEFS.ESM_IMPORT_META_URL ??
        // use `document.currentScript.src` if available
        (typeof document !== 'undefined' ? (document.currentScript as HTMLScriptElement)?.src :
                                           // use `self.location.href` if available
                                           (typeof self !== 'undefined' ? self.location?.href : undefined));
```

Maybe we can try something like:

```ts
const isBrowser = typeof window !== 'undefined' && typeof document !== 'undefined';
const isServiceWorker = typeof self !== 'undefined' && self instanceof ServiceWorkerGlobalScope;

export const scriptSrc =
    // if Node.js, return undefined
    isNode ? undefined :
    // if It's ESM, use import.meta.url
    (typeof import.meta !== 'undefined' && import.meta.url ? import.meta.url :
    // use `document.currentScript.src` if in browser environment
    (isBrowser ? document.currentScript.src :
    // use `self.location.href` if in service worker environment
    (isServiceWorker ? self.location.href : undefined)));
```

**Comment by ggaabe** [comment]
@fs-eire EDIT: Nvm the comment I just deleted, that error was because I didn't set the webpack `target` to `webworker`.

However, I'm getting a new error now (progress!):

`Error: no available backend found. ERR: [webgpu] RuntimeError: null function or function signature mismatch`

**Comment by ggaabe** [comment]
Update: Found the error is happening in here:

      if (!isInitializing) {
        backendInfo.initPromise = backendInfo.backend.init(backendName);
      }
      await backendInfo.initPromise;

For some reason the webgpu backend.init promise is rejecting due to the `null function or function signature mismatch` error. This is much further along than we were before though.

**Comment by fs-eire** [comment]
> Update: Found the error is happening in here:
> 
> https://github.com/microsoft/onnxruntime/blob/fff68c3151b774d8a2e9290e96b9f707cd950216/js/common/lib/backend-impl.ts#L83-L86
> 
> For some reason the webgpu backend.init promise is rejecting due to the `null function or function signature mismatch` error. This is much further along than we were before though.

Could you share me the reproduce steps?

**Comment by ggaabe** [comment]
@fs-eire You'll need to run the webGPU setup in a chrome extension. 

1. You can use my code I just published here: https://github.com/ggaabe/extension
2. run `npm install`
3. run `npm run build`
4. open the chrome manage extensions
<img width="342" alt="Screenshot 2024-06-14 at 9 37 14 AM" src="https://github.com/microsoft/onnxruntime/assets/5441185/e9ac0f03-1379-49b4-b5d4-a1bfe1842da8">

5. load unpacked
<img width="442" alt="Screenshot 2024-06-14 at 9 37 52 AM" src="https://github.com/microsoft/onnxruntime/assets/5441185/73da6d0a-de1c-4e4f-9eaa-24a2936b886b">

6. select the `build` folder from the repo.
7.  open the `AI WebGPU Extension` extension
8. type some text in the text input. it will load Phi-3 mini and after finishing loading this error will occur

**Comment by fs-eire** [comment]
@ggaabe I did some debug on my box and made some fixes -

1. Changes to ONNXRuntime Web:

   #21073 is created to make sure the web assembly file can be loaded correctly when `env.wasm.wasmPaths` is not specified.

2. Changes to https://github.com/ggaabe/extension

   https://github.com/ggaabe/extension/pull/1 need to be made to the extension example, to make it load the model correctly. Please note:
   - The onnxruntime-web version need to be updated to consume changes from (1) (after it get merged and published for dev channel)
   - There are still errors in background.js, which looks like incorrect params passed to `tokenizer.apply_chat_template()`. However, the WebAssembly is initialized and the model loaded successfully.

3. Other issues:
   - Transformerjs overrides `env.wasm.wasmPaths` to a CDN URL internally. At least for this example, we don't want this behavior so we need to reset it to `undefined` to keep the default behavior.
   - Multi-threaded CPU EP is not supported because `Worker` is not accessible in service worker. Issue tracking: https://github.com/whatwg/html/issues/8362

**Comment by ggaabe** [comment]
Awesome, thank you for your thoroughness in explaining this and tackling this head on. Is there a dev channel version I can test out?

**Comment by fs-eire** [comment]
Not yet. Will update in this thread once it is ready.

**Comment by ggaabe** [comment]
sorry to bug; is there any dev build number? wasn't sure how often a release runs

**Comment by fs-eire** [comment]
> sorry to bug; is there any dev build number? wasn't sure how often a release runs

Please try 1.19.0-dev.20240621-69d522f4e9

**Comment by ggaabe** [comment]
Interesting, this gives a couple of new errors:
1. ` [W:onnxruntime:, session_state.cc:1166 VerifyEachNodeIsAssignedToAnEp] Some nodes were not assigned to the preferred execution providers which may or may not have an negative impact on performance. e.g. ORT explicitly assigns shape related ops to CPU to improve perf.`
2. `[W:onnxruntime:, session_state.cc:1168 VerifyEachNodeIsAssignedToAnEp] Rerunning with verbose output on a non-minimal build will show node assignments.`

Did you see any errors like this?

**Comment by ggaabe** [comment]
@fs-eire I'm getting one new error:

```
ort.webgpu.bundle.min.mjs:6 Uncaught (in promise) Error: The data is not on CPU. Use `getData()` to download GPU data to CPU, or use `texture` or `gpuBuffer` property to access the GPU data directly.
    at get data (ort.webgpu.bundle.min.mjs:6:13062)
    at get data (tensor.js:62:1)
```

I pushed the code changes, to reproduce, just type 1 letter in the text input and wait

**Comment by nickl1234567** [comment]
Hey, I also need this. I am struggling with importing this version. So far I have been importing ONNX using
`import * as ort from "https://cdn.jsdelivr.net/npm/onnxruntime-web/dist/esm/ort.webgpu.min.js"`.
However, when I change to `import * as ort from "https://cdn.jsdelivr.net/npm/onnxruntime-web@1.19.0-dev.20240621-69d522f4e9/dist/esm/ort.webgpu.min.js"` it seems not to have an `.../esm/` folder. Do you know why that is and how to import it then?

**Comment by fs-eire** [comment]
> Hey, I also need this. I am struggling with importing this version. So far I have been importing ONNX using `import * as ort from "https://cdn.jsdelivr.net/npm/onnxruntime-web/dist/esm/ort.webgpu.min.js"`. However, when I change to `import * as ort from "https://cdn.jsdelivr.net/npm/onnxruntime-web@1.19.0-dev.20240621-69d522f4e9/dist/esm/ort.webgpu.min.js"` it seems not to have an `.../esm/` folder. Do you know why that is and how to import it then?

just replace `.../esm/ort.webgpu.min.js` to `.../ort.webgpu.min.mjs` should work. If you are also using service worker, use `ort.webgpu.bundle.min.mjs` instead of `ort.webgpu.min.mjs`.

**Comment by fs-eire** [comment]
> @fs-eire I'm getting one new error:
> 
> ```
> ort.webgpu.bundle.min.mjs:6 Uncaught (in promise) Error: The data is not on CPU. Use `getData()` to download GPU data to CPU, or use `texture` or `gpuBuffer` property to access the GPU data directly.
>     at get data (ort.webgpu.bundle.min.mjs:6:13062)
>     at get data (tensor.js:62:1)
> ```
> 
> I pushed the code changes to my repo and fixed the call to the tokenizer. To reproduce, just type 1 letter in the chrome extension’s text input and wait

This may be a problem of transformerjs. Could you try whether this problem happen in a normal page? If so, can report the issue to transformerjs. If it's only happening in service worker, I can take a closer look

**Comment by kyr0** [comment]
@fs-eire I can verify that using `1.19.0-dev.20240621-69d522f4e9` loading a model using `webgpu` in a service worker works - even in a web extension. The necessary code is:

```ts
import * as ONNX_WEBGPU from "onnxruntime-web/webgpu";

// any Blob that contains a valid ORT model would work
// I'm using Xenova/multilingual-e5-small/onnx/model_quantized.with_runtime_opt.ort
const buffer = await mlModel.blob.arrayBuffer();  
  
const sessionwebGpu = await ONNX_WEBGPU.InferenceSession.create(buffer, {
  executionProviders: ["webgpu"],
});
console.log("Loading embedding model using sessionwebGpu", sessionwebGpu);
```

Results in a successful execution, yay! 💯 :)

<img width="788" alt="Bildschirmfoto 2024-07-06 um 19 41 18" src="https://github.com/microsoft/onnxruntime/assets/454817/935c943e-d68d-4684-8dae-15f11d472dc2">

I think we can ignore the warning, printed as an error, as the session loads.

WebAssembly would work in a Service Worker. Just because Service Workers are limited in their ability to load external resources such as WASM runtime files as `Blob` or `ArrayBuffer` doesn't mean you can't get such data transferred **into** the Service Worker context. In fact, you can transfer Gigabytes __instantly__ using `MessageChannel` and the concept of  [Transferable objects](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Transferable_objects).

Passing down a `Blob/ArrayBuffer` from a content script to a background worker/service worker even works, standard-compliant, with Web Extensions, as I demonstrate here: https://github.com/w3c/webextensions/issues/293#issuecomment-2211770512

It's even much simpler for non-Web-Extension use cases as you simply only use the `self.onmessage` API in a service worker to receive a `MessageChannel` object and via a `port` of it, receive the `Blob` or `ArrayBuffer`. 

So... I think there is no need to always explicitly dynamically `import()` the WASM runtime in ONNX runtime code. It probably should be the default, but it would be amazing if there was a way to pass the `ArrayBuffer` of the WASM runtime directly via the options, by reference. Any developer could then load the WASM runtime data in various ways that fit their scenario and pass it down. 

Proposal:
```ts
  const buffer = await mlModel.blob.arrayBuffer();
  const wasmRuntime = await wasmRuntime.blob.arrayBuffer();
  const sessionWasm = await ONNX_WASM.InferenceSession.create(buffer, {
    executionProviders: ["wasm"],
    // if it is passed, the dynamic import code is not executed, but the ArrayBuffer of this reference is used
    wasmRuntime
  });
```

Currently, when I run:

```ts
const sessionWasm = await ONNX_WASM.InferenceSession.create(buffer, {
  executionProviders: ["wasm"],
});
console.log("Loading embedding model using sessionWasm", sessionWasm);
```

Result:
<img width="789" alt="Bildschirmfoto 2024-07-06 um 19 57 46" src="https://github.com/microsoft/onnxruntime/assets/454817/52091ffc-a495-4228-9999-0d077b5ba6b7">

Thank you for your help!

**Comment by ChTiSh** [comment]
I can confirm Web GPU is working for my little chrome extension app as well, but I'm having a problem disabling the warning.

**Comment by kyr0** [comment]
> I can confirm Web GPU is working for my little chrome extension app as well, but I'm having a problem disabling the warning.

You can numb it using a brittle monkey patch...

```ts
const originalConsole = self.console;

self.console.error = () => {}

const sessionwebGpu = await ONNX_WEBGPU.InferenceSession.create(buffer, {
  executionProviders: ["webgpu"],
});
console.log("Loading embedding model using sessionwebGpu", sessionwebGpu);

self.console.error = originalConsole.error;
```

**Comment by ChTiSh** [comment]
Thank you so much!!!!! The whole time I was trying to change the ort log
severity, now it's fast and beautiful!!!!

**Comment by kyr0** [comment]
@ChTiSh You're welcome 🫶 Always happy to help :)

**Comment by kyr0** [comment]
> > @fs-eire I'm getting one new error:
> > ```
> > ort.webgpu.bundle.min.mjs:6 Uncaught (in promise) Error: The data is not on CPU. Use `getData()` to download GPU data to CPU, or use `texture` or `gpuBuffer` property to access the GPU data directly.
> >     at get data (ort.webgpu.bundle.min.mjs:6:13062)
> >     at get data (tensor.js:62:1)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I pushed the code changes to my repo and fixed the call to the tokenizer. To reproduce, just type 1 letter in the chrome extension’s text input and wait
> 
> This may be a problem of transformerjs. Could you try whether this problem happen in a normal page? If so, can report the issue to transformerjs. If it's only happening in service worker, I can take a closer look

Did data structures change for the `Tensor` class? Specifically `dataLocation`vs `location`? And if so, did it change consistently? I'm facing issues with `data` being `undefined` but `cpuData` being set (tokenizer result). But when I pass the data down to a BERT model, `onnxruntime-web` seems to expect a different data structure and checks `location` and `data`. Am I missing something or has something changed? Could this lead to downstream issues of code checking for the `location` and `data` properties mistakenly believing the data isn't there or not in the right place? I linked a downstream issue..

**Comment by kyr0** [comment]
@fs-eire Is there any way how I could help pushing this forward? Thank you :)

**Comment by fs-eire** [comment]
@kyr0 thank you a lot for your willing to help. I am currently in vacation but I will pick up this thread when I am back by end of this month.

**Comment by kyr0** [comment]
@fs-eire Oh, I didn't mean to disturb you on vacation. Please enjoy, relax and have a lot of fun!

**Comment by asadm** [comment]
@kyr0 do you by any chance have a fork that I can try with the custom `env.wasm.instantiateWasm` addition?

I would like to try your workaround since I am stuck in a similar spot (.wasm needs to be imported in userland code)

**Comment by kyr0** [comment]
@asadm Well, I do have a solution for you to PoC if it would work, but I don't have a fork/PR yet. It's a bit messy, but I'll explain. If your code or a library that uses `onnxruntime-web` is importing the library, the module resolution algorithm will discover `node_modules/onnxruntime-web/package.json` and check for the `exports` defined in order to decide on which actual file to import. Your code, or the library code that your code is using, may have an import statement like this:

```ts
import { ... } from "onnxruntime-web/webgpu"
```

I decided, just to try, to simply change the code of `onnxruntime-web` on the fly. But the library only exports minified code by default. So I changed it's package.json exports to point to the non-minified code files instead:


```json
{
    ".": {
      "node": {
        "import": "./dist/ort.mjs",
        "require": "./dist/ort.js"
      },
      "import": "./dist/ort.mjs",
      "require": "./dist/ort.js",
      "types": "./types.d.ts"
    },
    "./all": {
      "node": null,
      "import": "./dist/ort.all.bundle.min.mjs",
      "require": "./dist/ort.all.min.js",
      "types": "./types.d.ts"
    },
    "./wasm": {
      "node": null,
      "import": "./dist/ort.wasm.mjs",
      "require": "./dist/ort.wasm.js",
      "types": "./types.d.ts"
    },
    "./webgl": {
      "node": null,
      "import": "./dist/ort.webgl.min.mjs",
      "require": "./dist/ort.webgl.min.js",
      "types": "./types.d.ts"
    },
    "./webgpu": {
      "node": null,
      "import": "./dist/ort.webgpu.mjs",
      "require": "./dist/ort.webgpu.js",
      "types": "./types.d.ts"
    },
    "./training": {
      "node": null,
      "import": "./dist/ort.training.wasm.min.mjs",
      "require": "./dist/ort.training.wasm.min.js",
      "types": "./types.d.ts"
    }
  },
```

If you do that change in your local filesystem in your project, your build process will now point to those files, no matter what your build system looks like. 

The next thing was to make `onnxruntime-web/webgpu` use my function if it is defined on `env`. There is 1 LoC where the call happens and it can access `env` in that scope. So I changed that code locally. Depending on what your code or the library you are using, is importing from `onnxruntime-web`, it might be one of those files. Or you change all of them, just to make sure.
`node_modules/onnxruntime-web/dist/ort.mjs` (line 24244), or `node_modules/onnxruntime-web/dist/ort.wasm.mjs` (line 1763):
Please note that the `webgpu` backend has no WASM import logic.

BEFORE:
<img width="1345" alt="Bildschirmfoto 2024-07-20 um 17 44 26" src="https://github.com/user-attachments/assets/e81de256-b6f6-4539-a708-bb6a5a385aa1">

AFTER:
<img width="1217" alt="Bildschirmfoto 2024-07-20 um 18 03 58" src="https://github.com/user-attachments/assets/c120f2e4-31b1-40c3-a14a-52a5f1a4e582">

I automated this process in a little NPM postinstall script, as I didn't want to spend much time on figuring all the build processes of `onnxruntime-web` yet.

```js
.replace(/importWasmModule\(/g, '(typeof env.importWasmModule === "function" ? env.importWasmModule : importWasmModule)(')
```
I know, I know. Hacky.. but pragmatic. You will loose your changes each time you re-install your dependencies.

But, after all, you can now simply assign the function to the `env` and it will be called:

```ts

// @ts-ignore
import getModule from "./node_modules/onnxruntime-web/dist/ort-wasm-simd-threaded.jsep";

// you may need to copy this file and the WASM file into a folder so that the loader can fetch() it well
./node_modules/onnxruntime-web/dist/ort-wasm-simd-threaded.wasm

// this is a working example in my project - it loads just fine now
  env.backends.onnx.importWasmModule = async (
    mjsPathOverride: string,
    wasmPrefixOverride: string,
    threading: boolean,
  ) => {
    console.log(
      "importWasmModule",
      mjsPathOverride,
      wasmPrefixOverride,
      threading,
    );

    return [
      undefined,
      async (moduleArgs = {}) => {
        console.log("moduleArgs", moduleArgs);
        return await getModule(moduleArgs);
      },
    ];
  };
```

My proposal would be, just to change this one line of code in this project to allow for optional Inversion of Control @fs-eire -- it could be documented with my example code. This would probably all issues regarding "user-land based WASM loading". 

Okay, I made a PR for that: https://github.com/microsoft/onnxruntime/pull/21430

**Comment by asadm** [comment]
@kyr0 that is amazing! I was also hacking around with unminified bundle (don't want to rebuild from source etc). 

Thank you so much for detailed solution, can't wait to try this when I get home!

**Comment by kyr0** [comment]
@asadm You're welcome. No worries, we'll get this to work just fine, also for you :) Here's an impression from by background worker with the monkey patch applied. It's working, even through the `@xenova/transformers.js` abstraction layer in between:

<img width="1527" alt="350204823-86f101f4-5213-4b5a-bc94-4c6bbcce047b" src="https://github.com/user-attachments/assets/8e01d588-27a1-4a61-a0fe-e5d44da80d3a">

Might take a while though, until downstream projects will adopt the new version. But once the maintainers here deploy a new version including my PR, we will be able to just override the `onnxruntime-web` dependencies using  package [`overrides`](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides)

**Comment by kyr0** [comment]
I forgot to mention that you need to also alter the emscripten-generated WASM loader code too, or import a different build. emscripten generates the pthread variant with this code in the WASM runtime loader:

```ts
if (isNode) isPthread = (await import('worker_threads')).workerData === 'em-pthread';
```

For my PR, I should probably figure out a way to instruct emscripten to not generate code for Node.js, aka build a second variant of the runtime for Workers that has Node support explicitly disabled. When the runtime is worker, the default implementation could import that one instead. In special cases where the custom loader would be implemented by users, they could also then fallback to this variant.

Why? As stated in this conversation earlier: Top-level await isn't supported in Worker environments.

<img width="724" alt="Bildschirmfoto 2024-07-20 um 19 04 52" src="https://github.com/user-attachments/assets/dc2d23cc-4c32-4af7-8636-41ea14317173">

**Comment by kyr0** [comment]
@asadm https://github.com/kyr0/easy-embeddings demonstrates the whole process; I automated the monkey-patching for the moment... https://github.com/kyr0/easy-embeddings/blob/main/scripts/setup-transformers.ts

**Comment by fs-eire** [comment]
I created #21534, which is a replacement of #21430:
- exposing the `instantiateWasm` directly may be not a good idea because this requires user to understand details of how WebAssembly works. I think this is unnecessary. To fulfill the requirement, allowing user to set an ArrayBuffer of the .wasm file should be good enough.
- the `import() is disallowed ...` error is already fixed after version `1.19.0-dev.20240621-69d522f4e9`. (only work for ESM. UMD will not work)
- with this fix in ort-web, a few things are no longer need:
  - hacks on package.json and bundle
  - the top-level await is already taking care of by the build script of ort-web.

**Comment by kyr0** [comment]
@fs-eire Sounds fair, and thank you for your work on this. Is there a new dev release that contains #21534 and that I could use to test, maybe? I'd surely only use ESM, so there shouldn't be an issue with that.

**Comment by fs-eire** [comment]
1.19.0-dev.20240801-4b8f6dcbb6 includes the change.

**Comment by lucasgelfond** [comment]
To clarify, is the best way to go about running transformers.js with WebGPU for the onnxruntime to monkeypatch the package to make the necessary wasm stuff load in each service worker, a la @kyr0's `easy-embeddings`? (Having some issues with that workflow, see https://github.com/kyr0/easy-embeddings/issues/1)

Has anyone had luck / have tips for just running the v3 branch of Transformers.js? Or, maybe more precisely — do we know how something like [Segment Anything WebGPU](https://huggingface.co/spaces/Xenova/segment-anything-webgpu), which Xenova has in an HF Space, is working? Seems like there's been some official solution here but I can't find it documented / implemented well.

**Comment by fs-eire** [comment]
> To clarify, is the best way to go about running transformers.js with WebGPU for the onnxruntime to monkeypatch the package to make the necessary wasm stuff load in each service worker, a la @kyr0's `easy-embeddings`? (Having some issues with that workflow, see [kyr0/easy-embeddings#1](https://github.com/kyr0/easy-embeddings/issues/1))
> 
> Has anyone had luck / have tips for just running the v3 branch of Transformers.js? Or, maybe more precisely — do we know how something like [Segment Anything WebGPU](https://huggingface.co/spaces/Xenova/segment-anything-webgpu), which Xenova has in an HF Space, is working? Seems like there's been some official solution here but I can't find it documented / implemented well.

I am working with Transformer.js to make v3 branch compatible with latest module system. This is one of the merged changes: https://github.com/xenova/transformers.js/pull/864. You probably need to use some workaround for now, but (hopefully) eventually you should be able to use it out of box.

**Comment by kyr0** [comment]
@lucasgelfond  Now that the new updates from @fs-eire are in place, I'm probably able to streamline the workaround. I'll have a look soon, but as I'm on vacation right now, I cannot give an ETA, unfortunately.

**Comment by lucasgelfond** [comment]
Thank you @fs-eire and @kyr0 ! No huge rush on my end, ended up getting inference working on WebGPU just on vanilla onnxruntime, will share results in a bit!

**Comment by lucasgelfond** [comment]
React implementation works although code is super sloppy! Skipping Transformers.js entirely: https://github.com/lucasgelfond?tab=repositories

Now rebuilding in Svelte, just FYI, Vite imports for the WebGPU bit here don't really work, here's error:

(Although, very easily fixable with the method @fs-eire mentions above, thanks again!)

Just used:

  ```import { InferenceSession, Tensor as ONNX_TENSOR } from 'https://cdn.jsdelivr.net/npm/onnxruntime-web@1.20.0-dev.20240810-6ae7e02d34/dist/ort.webgpu.min.js';```

Anyways, error in case anyone else is searching etc:


```


  VITE v5.4.0  ready in 400 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
  ➜  press h + enter to show help
Error:   Failed to scan for dependencies from entries:
  /Users/focus/Projects/---/webgpu-sam2/frontend/src/routes/+page.svelte

  ✘ [ERROR] Missing "./webgpu/index.js" specifier in "onnxruntime-web" package [plugin vite:dep-scan]

    script:/Users/focus/Projects/---/webgpu-sam2/frontend/src/encoder.svelte?id=0:6:31:
      6 │   import * as ONNX_WEBGPU from "onnxruntime-web/webgpu";
        ╵                                ~~~~~~~~~~~~~~~~~~~~~~~~

  This error came from the "onResolve" callback registered here:

    node_modules/esbuild/lib/main.js:1150:20:
      1150 │       let promise = setup({
           ╵                     ^
    at setup (file:///Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:49654:13)
    at handlePlugins (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:1150:21)
    at buildOrContextImpl (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:873:5)
    at Object.buildOrContext (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:699:5)
    at /Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:2032:68
    at new Promise (<anonymous>)
    at Object.context (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:2032:27)
    at Object.context (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:1874:58)
    at prepareEsbuildScanner (file:///Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/vite/dist/node/chunks/dep-NjL7WTE1.js:49451:24)


    at failureErrorWithLog (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:1472:15)
    at /Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:945:25
    at runOnEndCallbacks (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:1315:45)
    at buildResponseToResult (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:943:7)
    at /Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:955:9
    at new Promise (<anonymous>)
    at requestCallbacks.on-end (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:954:54)
    at handleRequest (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:647:17)
    at handleIncomingPacket (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:672:7)
    at Socket.readFromStdout (/Users/focus/Projects/---/webgpu-sam2/frontend/node_modules/esbuild/lib/main.js:600:7)
3:23:12 PM [vite] Pre-transform error: /Users/focus/Projects/---/webgpu-sam2/frontend/src/encoder.svelte:8:12 Unexpected token
3:23:12 PM [vite] Error when evaluating SSR module /src/routes/+page.svelte:


```

**Comment by lucasgelfond** [comment]
Has anyone tried getting these imports working in Vite/other bundlers? When I try the classic:

``import * as ONNX_WEBGPU from 'onnxruntime/webgpu``

(which works in create-react-app), Vite says:

```
Error:   Failed to scan for dependencies from entries:
  /Users/focus/Projects/---/webgpu-sam2/frontend/src/routes/+page.svelte

  ✘ [ERROR] Missing "./webgpu/index.js" specifier in "onnxruntime-web" package [plugin vite:dep-scan]

```

I disabled SSR in Svelte but still seemingly no luck.

Anyways, I tried importing from url, a la

``
  import { InferenceSession, Tensor as ONNX_TENSOR } from 'https://cdn.jsdelivr.net/npm/onnxruntime-web@1.20.0-dev.20240810-6ae7e02d34/dist/ort.webgpu.min.js';
``

which Vite also doesn't like

```
3:34:51 PM [vite] Error when evaluating SSR module /src/encoder.svelte: failed to import "https://cdn.jsdelivr.net/npm/onnxruntime-web@1.20.0-dev.20240810-6ae7e02d34/dist/ort.webgpu.min.js"
|- Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data, and node are supported by the default ESM loader. 
```

I tried manually downloading the files with CURL, where I got an error about the lack of source map, so, I also downloaded .min.js.map. When I run it now, this works, but I get back to the original error in the thread about unavailable backends:

```
Error: no available backend found. ERR: [webgpu] TypeError: Failed to fetch dynamically imported module: http://localhost:5173/src/ort-wasm-simd-threaded.jsep.mjs
```

I figured it might work to just import directly, so I also tried:

`import * as ONNX_WEBGPU from 'onnxruntime-web/dist/ort.webgpu.min.mjs';`

but then I got
``
4:07:38 PM [vite] Internal server error: Missing "./dist/ort.webgpu.min.mjs" specifier in "onnxruntime-web" package
``


Anyone have ideas of how to handle? Happy to add more verbose error messages for any of the stuff above.

**Comment by fs-eire** [comment]
> Has anyone tried getting these imports working in Vite/other bundlers? When I try the classic:
> 
> `import * as ONNX_WEBGPU from 'onnxruntime/webgpu`
> 
> (which works in create-react-app), Vite says:
> 
> ```
> Error:   Failed to scan for dependencies from entries:
>   /Users/focus/Projects/---/webgpu-sam2/frontend/src/routes/+page.svelte
> 
>   ✘ [ERROR] Missing "./webgpu/index.js" specifier in "onnxruntime-web" package [plugin vite:dep-scan]
> ```
> 
> Anyways, I tried importing from url, a la
> 
> `import { InferenceSession, Tensor as ONNX_TENSOR } from 'https://cdn.jsdelivr.net/npm/onnxruntime-web@1.20.0-dev.20240810-6ae7e02d34/dist/ort.webgpu.min.js';`
> 
> which Vite also doesn't like
> 
> ```
> 3:34:51 PM [vite] Error when evaluating SSR module /src/encoder.svelte: failed to import "https://cdn.jsdelivr.net/npm/onnxruntime-web@1.20.0-dev.20240810-6ae7e02d34/dist/ort.webgpu.min.js"
> |- Error [ERR_UNSUPPORTED_ESM_URL_SCHEME]: Only URLs with a scheme in: file, data, and node are supported by the default ESM loader. 
> ```
> 
> I disabled SSR in Svelte but still seemingly no luck/change.
> 
> I tried manually downloading the files with CURL, where I got an error about the lack of source map, so, I also downloaded .min.js.map. When I run it now, this works, but I get back to the original error in the thread about unavailable backends:
> 
> ```
> Error: no available backend found. ERR: [webgpu] TypeError: Failed to fetch dynamically imported module: http://localhost:5173/src/ort-wasm-simd-threaded.jsep.mjs
> ```
> 
> I figured it might work to just import directly, so I also tried:
> 
> `import * as ONNX_WEBGPU from 'onnxruntime-web/dist/ort.webgpu.min.mjs';`
> 
> but then I got `4:07:38 PM [vite] Internal server error: Missing "./dist/ort.webgpu.min.mjs" specifier in "onnxruntime-web" package`
> 
> Anyone have ideas of how to handle? Happy to add more verbose error messages for any of the stuff above.

Could you share me a repo that I can reproduce the issue? I will take a look.

**Comment by lucasgelfond** [comment]
(Non-working Vite stuff is now in previous commits because I shipped this!)

If you want to play with it, SAM-2 working totally in WebGPU! [Here's](https://webgpu-sam2.lucasgelfond.online/) a live demo link, thanks all for the help!

![demo](https://github.com/user-attachments/assets/e8f41540-0584-4b6f-a9e7-4efbdfa48734)

**Comment by Eldow** [comment]
👋 Thank you @fs-eire ! I tried using `1.19.0-dev.20240801-4b8f6dcbb6` inside of a chrome mv3 extension and it worked right away with the webgpu backend, however I'm more interested in using the wasm backend for running a simple decision forest as it doesn't include jsep and makes the overall bundle 10MB lighter. I wondered if there were any plans to support it too in a near future ?

**Comment by Eldow** [comment]
It works indeed 😵 I tried doing `import "onnxruntime-web/wasm"` at first and this one still returned `ERR: [wasm] TypeError: import() is disallowed on ServiceWorkerGlobalScope by the HTML specification`

**Comment by juliankolbe** [comment]
So im still having this issue in 1.19.2.

This is in the context of a chrome extension, mv3, `webgpu` works fine. However `cpu/wasm` (as far as i understand they are aliases) does not. Ive been through the code and to me looks like the dynamic import is always happening and cant be skipped.

This:
https://github.com/microsoft/onnxruntime/blob/e91ff9438bd368300b7d6d95aabfffe8cb9547b6/js/web/lib/wasm/wasm-factory.ts#L119

Calling this:
https://github.com/microsoft/onnxruntime/blob/e91ff9438bd368300b7d6d95aabfffe8cb9547b6/js/web/lib/wasm/wasm-utils-import.ts#L186

Seems to lead to this:

 `ERR: [cpu] TypeError: import() is disallowed on ServiceWorkerGlobalScope by the HTML specification`
 
 I realize the poster above me is running the same setup and has it working, but Im really not sure what to do differently.
 
Using code from the test file, ive tried replicating it like so, but this doesnt seem to work:

```
const binaryURL = ONNX_WASM_CDN_URL + "ort-wasm-simd-threaded.wasm";
const response = await fetch(binaryURL)
const binary = await response.arrayBuffer();
ort.env.wasm.wasmBinary = binary;

const session = await ort.InferenceSession.create(modelPath, {
  executionProviders: ["cpu"],
});
```

**Comment by fs-eire** [comment]
> So im still having this issue in 1.19.2.
> 
> This is in the context of a chrome extension, mv3, `webgpu` works fine. However `cpu/wasm` (as far as i understand they are aliases) does not. Ive been through the code and to me looks like the dynamic import is always happening and cant be skipped.
> 
> This:
> 
> https://github.com/microsoft/onnxruntime/blob/e91ff9438bd368300b7d6d95aabfffe8cb9547b6/js/web/lib/wasm/wasm-factory.ts#L119
> 
> Calling this:
> 
> https://github.com/microsoft/onnxruntime/blob/e91ff9438bd368300b7d6d95aabfffe8cb9547b6/js/web/lib/wasm/wasm-utils-import.ts#L186
> 
> Seems to lead to this:
> 
> `ERR: [cpu] TypeError: import() is disallowed on ServiceWorkerGlobalScope by the HTML specification`
> 
> I realize the poster above me is running the same setup and has it working, but Im really not sure what to do differently.
> 
> Using code from the test file, ive tried replicating it like so, but this doesnt seem to work:
> 
> ```
> import ort from "onnxruntime-web";
> 
> const binaryURL = ONNX_WASM_CDN_URL + "ort-wasm-simd-threaded.wasm";
> const response = await fetch(binaryURL)
> const binary = await response.arrayBuffer();
> ort.env.wasm.wasmBinary = binary;
> ort.env.wasm.numThreads = 1;
> 
> const session = await ort.InferenceSession.create(modelPath, {
>   executionProviders: ["cpu"],
> });
> ```

If you are using 1.19.2 and still ran into this error, it is probably because your bundler imports onnxruntime-web as UMD. please verify the following:

- check the version is 1.19.2 in <your_project_root>/node_modules/onnxruntime-web/package.json
- check if your bundler actually loads file <your_project_root>/node_modules/onnxruntime-web/dist/ort.bundle.min.mjs. You can backup this file, then put a few illegal characters inside and run your bundler to quickly check. if you bundler still builds, that means it's not picking this file.

**Comment by juliankolbe** [comment]
the bundler was indeed the issue, for anyone reading this: I was using vite 4 and it was prefering the browser field in the package.json which led to the wrong file. Switching to vite 5 solves that issue as you can change the order of fields, even though it will by default already prefer the exports field.

I have another issue now though: Now that the correct file has made it, I am getting this error:

 `Error: no available backend found. ERR: [cpu] ReferenceError: Worker is not defined`
 
 any ideas what that might be?

**Comment by fs-eire** [comment]
In my understanding, `Worker` is unavailable in service worker. I have no idea whether this is by design or not. However, if you don't use multi-thread (set `ort.env.wasm.numThreads = 1;`) and disable proxy (do not set `ort.env.wasm.proxy`), onnxruntime-web should not use `Worker`.

**Comment by Eldow** [comment]
I have a follow-up issue that I struggle to understand: when using onnxruntime-web in a mv3 extension service-worker bundled with webpack.

The symptoms are that the service worker will crash right away with this error:
```
Uncaught ReferenceError: document is not defined

...

/******/ 	/* webpack/runtime/jsonp chunk loading */
/******/ 	(() => {
/******/ 		__webpack_require__.b = document.baseURI || self.location.href;
```

Here is a small reproduction sandbox: 
https://codesandbox.io/p/devbox/cocky-hamilton-5m398f

(Run `yarn build` and download the dist folder to try it out in a web browser where you can load unpacked extensions - I tested in Chrome & firefox)

There is a way to make it work using `output.chunkLoading = false` in the webpack configuration but this is not ideal.
Could it be related to loading wasm with webpack: https://stackoverflow.com/a/71673305 ?

I am not sure if this requires a dedicated issue or if it's okay to follow up in this one, let me know @fs-eire

**Comment by fs-eire** [comment]
> I have a follow-up issue that I struggle to understand: when using onnxruntime-web in a mv3 extension service-worker bundled with webpack.
> 
> The symptoms are that the service worker will crash right away with this error:
> 
> ```
> Uncaught ReferenceError: document is not defined
> 
> ...
> 
> /******/ 	/* webpack/runtime/jsonp chunk loading */
> /******/ 	(() => {
> /******/ 		__webpack_require__.b = document.baseURI || self.location.href;
> ```
> 
> Here is a small reproduction sandbox: https://codesandbox.io/p/devbox/cocky-hamilton-5m398f
> 
> (Download the dist folder to try it out in a web browser where you can load unpacked extensions - I tested in Chrome & firefox)
> 
> There is a way to make it work using `output.chunkLoading = false` in the webpack configuration but this is not ideal. Could it be related to loading wasm with webpack: https://stackoverflow.com/a/71673305 ?
> 
> I am not sure if this requires a dedicated issue or if it's okay to follow up in this one, let me know [@fs-eire](https://github.com/fs-eire)

It looks like that you are using jsonp for chunk loading in webpack. I don't think this will work in a service worker.

I think the solution is to add `target: 'webworker',` like this:
```diff
 module.exports = {
   mode: "development",
   devtool: false,
+  target: 'webworker',
   entry: {

```

Then I can see another error saying that the model is invalid, which is expected:

```
[Background] Error loading ONNX model or running inference: Can't create a session. ERROR_CODE: 7, ERROR_MESSAGE: Failed to load model because protobuf parsing failed
```

**Comment by Eldow** [comment]
Thanks for looking into it ! 
I think this is what I will end up doing but I am wondering if webpack could have used the correct loader if emscripten generated a slightly different code to locate the wasm file without the fallback which relies on URL.

That could maybe be solved by generating a worker compatible only JS file through emscripten using the flag `ENVIRONMENT="worker"` ? (I am building from source using [the doc](https://onnxruntime.ai/docs/build/web.html) so I have control over the args passed to emscripten at that stage)

I will give it a try and let you know !

**Comment by xcaptain** [comment]
Same problem here, how should I solve the `no available backend` error in chrome extension background script?

**Comment by fs-eire** [comment]
Please refer to https://github.com/microsoft/onnxruntime/pull/20991#issuecomment-2622947140, which contains latest solution and explanation for using ORT-web in service worker.

**Comment by fs-eire** [comment]
> Thanks for looking into it ! I think this is what I will end up doing but I am wondering if webpack could have used the correct loader if emscripten generated a slightly different code to locate the wasm file without the case that relies on URL.
> 
> That could maybe be solved by generating a worker compatible only JS file through emscripten using the flag `ENVIRONMENT="worker"` ? (I am building from source using [the doc](https://onnxruntime.ai/docs/build/web.html) so I have control over the args passed to emscripten at that stage)
> 
> I will give it a try and let you know !
> 
> edit: I tried adding `--emscripten_settings ENVIRONMENT="worker"` to the build wasm command and faced new issues in the esbuild compilation afterwards regarding `Unexpected number of matches for minified "PThread.pthreads[thread].ref()"` - I'm stopping trying to do it this way in favor of the webpack config way.

I tried to use Emscripten to build multiple targets (https://github.com/emscripten-core/emscripten/issues/21899), but it turns out difficult to do. Currently the onnxruntime-web build process has a few post-processing to make them working for Web, WebWorker and Node.

It may also be a problem for how to tell the runtime or bundler to import the correct file, giving the predefined "conditions" in the package.json "exports" property do not contain anything that distinguish between web and webworker.

**Comment by Eldow** [comment]
I think you're right, since emscripten outputs a web compatible file and webpack by default uses the `web` target it ends up trying to use the `document` jsonp style loader. I'm not sure if there is a way to differentiate them so webpack can pick the right import strategy automatically without needing a hint with `target: 'webworker'` but maybe different exports for web & worker and different emscripten pipelines could do the job. I'm also not entirely sure onnxruntime-web would work in the background script of a MV2 extension (workers are only MV3) - but most browsers are compliant with MV3.

**Comment by juliankolbe** [comment]
@fs-eire Actually have a different question, so I managed to get `onnxruntime-web/webgpu` with wegpu running very well. The only issue I have is that it actually uses a mix of GPU and CPU, using all CPU cores and not fully utilising the GPU. 

The issue im having is that when i run a larger inference, it will block all other system tasks and browser ui etc. As its running max performance on all cores. 

My goal: Have 1 or 2 reservered cores, so basically supplying `numThreads` to the webgpu provider, such that it doesnt use all threads. 
I tried to use this setting `intraOpNumThreads` but it throws an error: 
```
onnxruntime-web_webgpu.js?v=a67bba03:1636 Uncaught (in promise) BindingError: _emval_take_value has unknown type 0xcf9a7
    at kn (onnxruntime-web_webgpu.js?v=a67bba03:1636:60)
    at ep (onnxruntime-web_webgpu.js?v=a67bba03:1766:33)
    at ort-wasm-simd-threaded.jsep.wasm:0x7fc2be
    at I.<computed> (onnxruntime-web_webgpu.js?v=a67bba03:2023:22)
    at ar (onnxruntime-web_webgpu.js?v=a67bba03:1270:39)
    at Yi (onnxruntime-web_webgpu.js?v=a67bba03:2052:129)
    at s (onnxruntime-web_webgpu.js?v=a67bba03:2055:13)
    at zo (onnxruntime-web_webgpu.js?v=a67bba03:920:20)
    at onnxruntime-web_webgpu.js?v=a67bba03:2003:17
```

Having that work would allow me to let the browser ui/rest of processes breathe while the longer inference which can take like a min.

**Comment by fs-eire** [comment]
> [@fs-eire](https://github.com/fs-eire) Actually have a different question, so I managed to get `onnxruntime-web/webgpu` with wegpu running very well. The only issue I have is that it actually uses a mix of GPU and CPU, using all CPU cores and not fully utilising the GPU.
> 
> The issue im having is that when i run a larger inference, it will block all other system tasks and browser ui etc. As its running max performance on all cores.
> 
> My goal: Have 1 or 2 reservered cores, so basically supplying `numThreads` to the webgpu provider, such that it doesnt use all threads. I tried to use this setting `intraOpNumThreads` but it throws an error:
> 
> ```
> onnxruntime-web_webgpu.js?v=a67bba03:1636 Uncaught (in promise) BindingError: _emval_take_value has unknown type 0xcf9a7
>     at kn (onnxruntime-web_webgpu.js?v=a67bba03:1636:60)
>     at ep (onnxruntime-web_webgpu.js?v=a67bba03:1766:33)
>     at ort-wasm-simd-threaded.jsep.wasm:0x7fc2be
>     at I.<computed> (onnxruntime-web_webgpu.js?v=a67bba03:2023:22)
>     at ar (onnxruntime-web_webgpu.js?v=a67bba03:1270:39)
>     at Yi (onnxruntime-web_webgpu.js?v=a67bba03:2052:129)
>     at s (onnxruntime-web_webgpu.js?v=a67bba03:2055:13)
>     at zo (onnxruntime-web_webgpu.js?v=a67bba03:920:20)
>     at onnxruntime-web_webgpu.js?v=a67bba03:2003:17
> ```
> 
> Having that work would allow me to let the browser ui/rest of processes breathe while the longer inference which can take like a min.
> 
> Edit: I am creating the session like this to get the error:
> 
> ```
> const session = await ort.InferenceSession.create(modelPath, {
>     executionProviders: ["webgpu"],
>     intraOpNumThreads: 14,
> })
> ```

The whole execution process is driven by CPU and the entry point of the inference execution is the wasm exported function `_OrtRun`. This means for heavy models if onnxruntime-web will occupy noticeable amount of CPU time and cause UI freeze.

If you are working on a product that has a UI responsive requirement, the solution is to use onnxruntime-web in a web worker and write communication logic between the worker and main thread.

Regarding the number of threads setting, for onnxruntime-web, [intraOpNumThreads](https://onnxruntime.ai/docs/api/js/interfaces/InferenceSession.SessionOptions.html#intraOpNumThreads) is not available ( you can see the document saying that it's only for onnxruntime-node and onnxruntime-react-native ). You should use `ort.env.wasm.numThreads` to set the number of threads instead. Also, this setting may be helpful for CPU(wasm) EP, but if you are using WebGPU, usually using multi-threads will not give you any performance benefits because we expect all heavy operators are calculated on GPU and CPU is only used for some very simple work like calculating tensor shapes. And for the reasons explained above, this setting does not help to imporve UI freeze. You need web worker for that.
